File: leetcode_python/Math/squirrel-simulation.py
# LeetCode 573. Squirrel Simulation

# There's a tree, a squirrel, and several nuts. Positions are represented by the cells in a 2D grid. Your goal is to find the minimal distance for the squirrel to collect all the nuts and put them under the tree one by one. The squirrel can only take at most one nut at one time and can move in four directions - up, down, left and right, to the adjacent cell. The distance is represented by the number of moves.

# Example 1:

# Input: 
# Height : 5
# Width : 7
# Tree position : [2,2]
# Squirrel : [4,4]
# Nuts : [[3,0], [2,5]]
# Output: 12
# Explanation:

# Note:

# All given positions won't overlap.
# The squirrel can take at most one nut at one time.
# The given positions of nuts have no order.
# Height and width are positive integers. 3 <= height * width <= 10,000.
# The given positions contain at least one nut, only one tree and one squirrel.



# V1 : dev 


# V2 
# http://bookshadow.com/weblog/2017/05/07/leetcode-squirrel-simulation/
class Solution(object):
    def minDistance(self, height, width, tree, squirrel, nuts):
        """
        :type height: int
        :type width: int
        :type tree: List[int]
        :type squirrel: List[int]
        :type nuts: List[List[int]]
        :rtype: int
        """
        manhattan = lambda p, q: abs(p[0] - q[0]) + abs(p[1] - q[1])
        total = 2 * sum(manhattan(tree, nut) for nut in nuts)
        return total + min(manhattan(squirrel, nut) - manhattan(tree, nut) for nut in nuts)


# V3 
# Time:  O(n)
# Space: O(1)
class Solution(object):
    def minDistance(self, height, width, tree, squirrel, nuts):
        """
        :type height: int
        :type width: int
        :type tree: List[int]
        :type squirrel: List[int]
        :type nuts: List[List[int]]
        :rtype: int
        """
        def distance(a, b):
            return abs(a[0] - b[0]) + abs(a[1] - b[1])

        result = 0
        d = float("inf")
        for nut in nuts:
            result += (distance(nut, tree) * 2)
            d = min(d, distance(nut, squirrel) - distance(nut, tree))
        return result + d

File: leetcode_java/src/main/java/LeetCodeJava/Math/SquirrelSimulation.java
package LeetCodeJava.Math;

// https://leetcode.com/problems/squirrel-simulation/

/**
 *  573. Squirrel Simulation
 *  Medium
 *
 *  You are given two integers height and width representing a garden of size
 *  height x width. You are also given:
 *    - an array tree where tree = [treer, treec] is the position of the tree in the garden,
 *    - an array squirrel where squirrel = [squirrelr, squirrelc] is the position of the
 *      squirrel in the garden,
 *    - and an array nuts where nuts[i] = [nutir, nutic] is the position of the ith nut.
 *
 *  The squirrel can manage at most one nut at a time and can move in four directions:
 *  up, down, left, and right, to the adjacent cell.
 *
 *  Return the minimal distance for the squirrel to collect all the nuts and put them under
 *  the tree one by one.
 *
 *  Example 1:
 *    Input: height = 5, width = 7, tree = [2,2], squirrel = [4,4], nuts = [[3,0],[2,5]]
 *    Output: 12
 *
 *  Example 2:
 *    Input: height = 1, width = 3, tree = [0,1], squirrel = [0,0], nuts = [[0,2]]
 *    Output: 3
 *
 *  Constraints:
 *    1 <= height, width <= 100
 *    tree.length == 2, squirrel.length == 2, nuts[i].length == 2
 *    1 <= nuts.length <= 5000
 *    All positions are inside the garden and are distinct.
 */
public class SquirrelSimulation {

    // V0
    // IDEA: every nut costs 2 * dist(tree, nut) except the FIRST one, which costs
    //       dist(squirrel, nut) + dist(tree, nut). So total = 2 * sum(dist(tree, nut))
    //       + min over nuts of (dist(squirrel, nut) - dist(tree, nut)).
    /**
     * time = O(n)
     * space = O(1)
     */
    public int minDistance(int height, int width, int[] tree, int[] squirrel, int[][] nuts) {
        int total = 0;
        int bestDelta = Integer.MAX_VALUE;
        for (int[] nut : nuts) {
            int toTree = dist(nut, tree);
            total += 2 * toTree;
            bestDelta = java.lang.Math.min(bestDelta, dist(nut, squirrel) - toTree);
        }
        return total + bestDelta;
    }

    private int dist(int[] a, int[] b) {
        return java.lang.Math.abs(a[0] - b[0]) + java.lang.Math.abs(a[1] - b[1]);
    }
}
